Ubuntu/devel

TheRealFalcon opened the pull request.

TheRealFalcon commented on debian/cloud-init.postinst:
```diff
@@ -327,6 +327,22 @@ fix_lp1889555() {
 
+change_cloud_init_output_permissions() {
+    # As a consequence of LP: #1918303
+    local oldver="$1" last_bad_ver="21.1-17-gc6726c2b-0ubuntu1"
```
Not sure I set this last bad version correctly. I went one back from the fix commit and determined the version from that.

OddBloke replied:
So I expect this would work, but normally I think we'd use the last version in which this was actually present in this release: `21.1-0ubuntu1` in this case. (I am also realising that cherry-picking this commit without modification won't work: I think we'll want to have the exact version for each release as `last_bad_ver`.)

Found via `rmadison cloud-init`, which produces:

```
 cloud-init | 0.6.3-0ubuntu1             | precise         | source, all
 cloud-init | 0.6.3-0ubuntu1.25          | precise-updates | source, all
 cloud-init | 0.7.5-0ubuntu1             | trusty          | source, all
 cloud-init | 0.7.5-0ubuntu1.18          | trusty-security | source, all
 cloud-init | 0.7.5-0ubuntu1.23          | trusty-updates  | source, all
 cloud-init | 0.7.7~bzr1212-0ubuntu1     | xenial          | source, all
 cloud-init | 18.2-14-g6d48d265-0ubuntu1 | bionic          | source, all
 cloud-init | 20.1-10-g71af48df-0ubuntu5 | focal           | source, all
 cloud-init | 20.3-15-g6d332e5c-0ubuntu1 | groovy          | source, all
 cloud-init | 20.4.1-0ubuntu1~16.04.1    | xenial-updates  | source, all
 cloud-init | 20.4.1-0ubuntu1~18.04.1    | bionic-updates  | source, all
 cloud-init | 20.4.1-0ubuntu1~20.04.1    | focal-updates   | source, all
 cloud-init | 20.4.1-0ubuntu1~20.10.1    | groovy-updates  | source, all
 cloud-init | 21.1-0ubuntu1              | hirsute         | source, all
```

OddBloke commented on debian/cloud-init.postinst:
```diff
@@ -327,6 +327,22 @@ fix_lp1889555() {
+        else
+            chown root $output_file
+        fi
+        chmod 740 $output_file
```
We shouldn't need this to be executable, so I think 640 would work?

OddBloke commented on debian/cloud-init.postinst:
```diff
@@ -327,6 +327,22 @@ fix_lp1889555() {
     db_set grub-pc/install_devices_empty "false"
 }
 
+change_cloud_init_output_permissions() {
```
A nit, but `output_log` makes it clearer what this is about when examining the list of all-the-things below: "cloud init output permissions" could mean a lot of things, "cloud init output log permissions" feels a little clearer.

OddBloke commented on debian/changelog:
```diff
@@ -1,5 +1,6 @@
-cloud-init (21.1-19-gbad84ad4-0ubuntu1) UNRELEASED; urgency=medium
-
+cloud-init (21.1-19-gbad84ad4-0ubuntu1) hirsute; urgency=medium
+  * d/cloud-init.postinst: debian/cloud-init.postinst: Change output log
```
I appreciate the thorough approach, but we can probably just mention the file path once. :grin:

OddBloke commented on debian/cloud-init.postinst:
```diff
@@ -327,6 +327,22 @@ fix_lp1889555() {
+
+    output_file="/var/log/cloud-init-output.log"
+    if [ -f "$output_file" ]; then
+        if $(grep -q adm /etc/group); then
```
No harm to leave it, but I _think_ we can assume the existence of `adm` on Ubuntu.

TheRealFalcon commented:
@OddBloke I believe I've addressed all of your comments. I also tested manually that the upgrade script works.

```
root@test:/tmp# ls -l /var/log
total 136
drwxr-xr-x  1 root      root                60 Mar 15 21:54 apt
-rw-r-----  1 syslog    adm                634 Mar 19 21:36 auth.log
-rw-rw----  1 root      utmp                 0 Mar 15 21:47 btmp
-rw-r--r--  1 root      root              5687 Mar 19 21:36 cloud-init-output.log
-rw-r--r--  1 syslog    adm              89550 Mar 19 21:36 cloud-init.log
drwxr-xr-x  1 root      root                 0 Jan 19 21:47 dist-upgrade
-rw-r--r--  1 root      adm                  0 Mar 19 21:36 dmesg
-rw-r--r--  1 root      root              8369 Mar 15 21:54 dpkg.log
drwxr-sr-x+ 1 root      systemd-journal     64 Mar 19 21:36 journal
drwxr-xr-x  1 landscape landscape            0 Apr 27  2020 landscape
-rw-rw-r--  1 root      utmp            292292 Mar 19 21:36 lastlog
drwx------  1 root      root                 0 Mar 19 21:36 private
-rw-r-----  1 syslog    adm              13450 Mar 19 21:36 syslog
drwxr-x---  1 root      adm                 64 Mar 19 21:36 unattended-upgrades
-rw-rw-r--  1 root      utmp              1536 Mar 19 21:36 wtmp
root@test:/tmp# dpkg -i cloud-init.deb 
(Reading database ... 31558 files and directories currently installed.)
Preparing to unpack cloud-init.deb ...
Unpacking cloud-init (21.1-19-gbad84ad4-0ubuntu1) over (20.4.1-0ubuntu1~20.04.1) ...
Setting up cloud-init (21.1-19-gbad84ad4-0ubuntu1) ...
Processing triggers for rsyslog (8.2001.0-1ubuntu1.1) ...
Processing triggers for man-db (2.9.1-1) ...
root@test:/tmp# ls -l /var/log
total 140
drwxr-xr-x  1 root      root                60 Mar 15 21:54 apt
-rw-r-----  1 syslog    adm                634 Mar 19 21:36 auth.log
-rw-rw----  1 root      utmp                 0 Mar 15 21:47 btmp
-rw-r-----  1 root      adm               5687 Mar 19 21:36 cloud-init-output.log
-rw-r--r--  1 syslog    adm              89550 Mar 19 21:36 cloud-init.log
drwxr-xr-x  1 root      root                 0 Jan 19 21:47 dist-upgrade
-rw-r--r--  1 root      adm                  0 Mar 19 21:36 dmesg
-rw-r--r--  1 root      root              9627 Mar 19 21:36 dpkg.log
drwxr-sr-x+ 1 root      systemd-journal     64 Mar 19 21:36 journal
drwxr-xr-x  1 landscape landscape            0 Apr 27  2020 landscape
-rw-rw-r--  1 root      utmp            292292 Mar 19 21:36 lastlog
drwx------  1 root      root                 0 Mar 19 21:36 private
-rw-r-----  1 syslog    adm              16564 Mar 19 21:36 syslog
drwxr-x---  1 root      adm                 64 Mar 19 21:36 unattended-upgrades
-rw-rw-r--  1 root      utmp              1536 Mar 19 21:36 wtmp
root@test:/tmp# chmod 007 /var/log/cloud-init-output.log 
root@test:/tmp# dpkg -i /tmp/cloud-init.deb 
(Reading database ... 31563 files and directories currently installed.)
Preparing to unpack /tmp/cloud-init.deb ...
Unpacking cloud-init (21.1-19-gbad84ad4-0ubuntu1) over (21.1-19-gbad84ad4-0ubuntu1) ...
Setting up cloud-init (21.1-19-gbad84ad4-0ubuntu1) ...
Processing triggers for rsyslog (8.2001.0-1ubuntu1.1) ...
Processing triggers for man-db (2.9.1-1) ...
root@test:/tmp# ls -l /var/log/cloud-init-output.log 
-------rwx 1 root adm 5687 Mar 19 21:36 /var/log/cloud-init-output.log
```

OddBloke commented on debian/cloud-init.postinst:
```diff
@@ -327,6 +327,22 @@ fix_lp1889555() {
+
+    output_file="/var/log/cloud-init-output.log"
+    if [ -f "$output_file" ]; then
+        if grep -q adm /etc/group; then
```
On building this, I'm seeing this lintian warning: https://lintian.debian.org/tags/maintainer-script-should-not-parse-etc-passwd-or-group.html

We should modify this to use `getent` per the recommendation there.

OddBloke reviewed (approved):
LGTM, thanks! I've confirmed that I see the same tree locally using our tooling (except for the postinst changes/changelog entry, of course), and that the package builds successfully. Uploading now.

OddBloke commented:
```
Trying to upload package to ubuntu
Checking signature on .changes
gpg: ../out/cloud-init_21.1-19-gbad84ad4-0ubuntu1_source.changes: Valid signature from EEE68776EDB13EA0
Checking signature on .dsc
gpg: ../out/cloud-init_21.1-19-gbad84ad4-0ubuntu1.dsc: Valid signature from EEE68776EDB13EA0
Uploading to ubuntu (via ftp to upload.ubuntu.com):
  Uploading cloud-init_21.1-19-gbad84ad4-0ubuntu1.dsc: done.
  Uploading cloud-init_21.1-19-gbad84ad4.orig.tar.gz: done.      
  Uploading cloud-init_21.1-19-gbad84ad4-0ubuntu1.debian.tar.xz: done.  
  Uploading cloud-init_21.1-19-gbad84ad4-0ubuntu1_source.buildinfo: done.
  Uploading cloud-init_21.1-19-gbad84ad4-0ubuntu1_source.changes: done.
Successfully uploaded packages.
```
